CSS Custom Properties Blog Post & Docs (#258)
tinacms/tinacms.org: content/blog/tinacms-styling-system.md

---
title: 'TinaCMS Styling System'
date: '2020-03-18T00:00:00.000Z'
author: Scott Byrne
---

## Writing an adaptable UI is hard.

When you're writing code for a quickly changing project, you want that code to be cheap and replaceable. [Styled Components](https://styled-components.com/) allow us to write styles right alongside the components they're used for. It lets us easily move components around — or completely replace them — with minimum effort.

> All these components — spread across many packages and files — need a single source of truth. Colors, padding sizes, shadows, fonts.. everything had to be consistent between TinaCMS components. Styled Components offered a powerful theming system that was a natural fit for the problem. We could provide theme context to any component looking to use common Tina styles.

Reviewer: ```suggestion
All these components — spread across many packages and files — needed a single source of truth. Colors, padding sizes, shadows, fonts.. everything had to be consistent between TinaCMS components. Styled Components offered a powerful theming system that was a natural fit for the problem. We could provide theme context to any component looking to use common Tina styles.
```

matches tense of the rest of the paragraph

> A major shift in our approach has been the ability to edit content directly inline where it's displayed. This meant that field specific UI would be displayed right on the user's website, not in an isolated TinaCMS container. We adapted our system to make this work, but the limitations of our approach became obvious. Even a simple button required type references to the styles package and a theme context provider. Theme changes — configured through gatsby-config.js — wouldn't take effect until the server restarted. We needed something more suited to the evolving TinaCMS project.

Reviewer: maybe a code example demonstrating how verbose it is to use a Tina button?

Author: Working on improving this para

Author: I couldn't really figure out a good example but I reworked this paragraph for clarity.

> A major shift in our approach has been the ability to edit content directly inline where it's displayed. This meant that field specific UI would be displayed right on the user's website, not in an isolated TinaCMS container. We adapted our system to make this work, but the limitations of our approach became obvious. Even a simple button required type references to the styles package and a theme context provider. Theme changes — configured through gatsby-config.js — wouldn't take effect until the server restarted. We needed something more suited to the evolving TinaCMS project.

Reviewer 2: > A major shift in our approach has been the ability to edit content
At first, I thought this was referring to our approach of using Styled Components, (and not our approach of Tina itself). Might be confusing for anyone reading this post without knowing a lot about Tina.

Author: I changed the first couple sentences:

> Initially, Tina UI was contained within an isolated sidebar inside an iframe. As we moved towards editing content inline with Tina, field specific UI would be displayed right on the user's website, not in an isolated TinaCMS container.

---
title: 'TinaCMS Styling System'
date: '2020-03-18T00:00:00.000Z'
author: Scott Byrne
---

## Writing an adaptable UI is hard.

When you're writing code for a quickly changing project, you want that code to be cheap and replaceable. [Styled Components](https://styled-components.com/) allow us to write styles right alongside the components they're used for. It lets us easily move components around — or completely replace them — with minimum effort.

> All these components — spread across many packages and files — needed a single source of truth. Colors, padding sizes, shadows, fonts.. everything had to be consistent between TinaCMS components. Styled Components offered a powerful theming system that was a natural fit for the problem. We could provide theme context to any component looking to use common Tina styles.

Reviewer: ```suggestion
All these components — spread across many packages and files — needed a single source of truth. Colors, padding sizes, shadows, fonts… everything had to be consistent between TinaCMS components. Styled Components offered a powerful theming system that was a natural fit for the problem. We could provide theme context to any component looking to use common Tina styles.
```

A major shift in our approach has been the ability to edit content directly inline where it's displayed. This meant that field specific UI would be displayed right on the user's website, not in an isolated TinaCMS container. We adapted our system to make this work, but the limitations of our approach became obvious. Even a simple button required type references to the styles package and a theme context provider. Theme changes — configured through gatsby-config.js — wouldn't take effect until the server restarted. We needed something more suited to the evolving TinaCMS project.

## Enter CSS custom properties.

> [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*) bring variables to the CSS masses! Instead of being compile time or dependant on JavaScript, they can be defined and used in plain CSS. CSS custom properties are declared and scoped to a selector, like `:root`. You can override existing properties by redeclaring them, either at the same scope or on a child. In the case of TinaCMS, that means you can easily customize and use our theme in your on project, without having to compile our styles.

Reviewer: ```suggestion
[CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*) bring variables to the CSS masses! Instead of being compile time or dependent on JavaScript, they can be defined and used in plain CSS. CSS custom properties are declared and scoped to a selector, like `:root`. You can override existing properties by redeclaring them, either at the same scope or on a child. In the case of TinaCMS, that means you can easily customize and use our theme in your own project, without having to compile our styles.
```

Reviewer: dependant -> dependent
in your on project -> in your own project

Here's an example where we override the primary color (blue by default) and use it in a custom button:

```css
:root {
  --tina-color-primary-light: #eb6337;
  --tina-color-primary: #ec4815;
  --tina-color-primary-dark: #dc4419;
}

.my-button {
  background-color: var(--tina-color-primary);
  border-radius: var(--tina-radius-big);

  :hover {
    background-color: var(--tina-color-primary-light);
  }
}
```

Reducing the complexity of our theme system makes it easier for us to adapt the UI to new features, but also makes it easier for those integrating TinaCMS with their project.

> [Take a look at the docs](https://tinacms.org/docs/cms/styles) for `@tinacms/styles` for a full reference of available properties.

## What's next?

> Styled Components are easy to use and generally a pleasure to work with. But unfortunately they require a runtime. While this isn't an issue in an app or website, it's problematic for a library. It's possible the library consumer is already using a different version of styled components. Or even a different css-in-js library that requires a separate runtime. Ideally TinaCMS styles would be framework agnostic with no runtime.

Reviewer 2: I don't know if this section says "What's next?", maybe more, "In Summary"?

Author: The problem is we haven't removed this 'runtime', so it's not really a summary of implemented changes.

Author: It's more like 'what's the next issue we're eyeing' in a way

Author: Updated to "What's the next challenge?"

Here's an example where we override the primary color (blue by default) and use it in a custom button:

```css
:root {
  --tina-color-primary-light: #eb6337;
  --tina-color-primary: #ec4815;
  --tina-color-primary-dark: #dc4419;
}

.my-button {
  background-color: var(--tina-color-primary);
  border-radius: var(--tina-radius-big);

  :hover {
    background-color: var(--tina-color-primary-light);
  }
}
```

Reducing the complexity of our theme system makes it easier for us to adapt the UI to new features, but also makes it easier for those integrating TinaCMS with their project.

> [Take a look at the docs](https://tinacms.org/docs/cms/styles) for `@tinacms/styles` for a full reference of available properties.

## What's next?

> Styled Components are easy to use and generally a pleasure to work with. But unfortunately they require a runtime. While this isn't an issue in an app or website, it's problematic for a library. It's possible the library consumer is already using a different version of styled components. Or even a different css-in-js library that requires a separate runtime. Ideally TinaCMS styles would be framework agnostic with no runtime.

Reviewer 2: > Styled Components are easy to use and generally a pleasure to work with. But unfortunately they require a runtime.

Maybe: Styled Components are easy to use and generally a pleasure to work with, but unfortunately it requires a runtime. 

Also, I'm not sure I fully understand what you mean here.

Author: I mean to say that a style runtime may sometimes be appropriate/beneficial for an app/website, but for a library to demand that a user downstream have this runtime isn't ideal. They may already have another active system in place. We want our styles to be 'dumb' and easily consumable regardless of what your app/website looks like.

Author: Last paragraph changed to:

> Styled Components are easy to use and generally a pleasure to work with, but unfortunately require a runtime. While this isn't an issue in an app or website, it's problematic for a library. It's possible the library consumer is already using a different version of styled components, or even a different css-in-js library that requires a separate runtime. Ideally TinaCMS styles would be framework agnostic with no runtime, consumable by any HTML page. Moving our theme system to CSS custom properties was the first step in aligning our style system with the needs of our project.

Reviewer 2: Cool, much better 👍
